dependencies don't render output (stderr?)

With a target
```
mypy:
      mypy .

lint: mypy
     black -t py310 .
```

The output does not show the dependency:
```
> must lint
black -t py310 --check .
would reformat test_costbasis.py
would reformat costbasis.py
would reformat __main__.py

Oh no! 💥 💔 💥
3 files would be reformatted, 1 file would be left unchanged.
thread 'main' panicked at 'assertion failed: result.status.success()', src/main.rs:38:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

makefile for reference:
```
> make -f mustfile lint
mypy .
costbasis.py:37: error: Argument 1 has incompatible type "Optional[T]"; expected "T"
costbasis.py:39: error: Incompatible types in assignment (expression has type "Purchase", target has type "T")
costbasis.py:43: error: Incompatible return value type (got "Optional[T]", expected "T")
costbasis.py:68: error: Incompatible return value type (got "Optional[Purchase]", expected "Purchase")
costbasis.py:77: error: Item "None" of "Optional[Purchase]" has no attribute "amount"
costbasis.py:78: error: Item "None" of "Optional[Purchase]" has no attribute "amount"
costbasis.py:79: error: Item "None" of "Optional[Purchase]" has no attribute "amount"
costbasis.py:80: error: Item "None" of "Optional[Purchase]" has no attribute "unit_cost"
costbasis.py:82: error: Incompatible types in assignment (expression has type "Union[float, Any]", variable has type "int")
costbasis.py:82: error: Item "None" of "Optional[Purchase]" has no attribute "unit_cost"
costbasis.py:87: error: Incompatible types in assignment (expression has type "float", variable has type "int")
__main__.py:6: error: Library stubs not installed for "dateutil" (or incompatible with Python 3.10)
__main__.py:6: note: Hint: "python3 -m pip install types-python-dateutil"
__main__.py:6: note: (or run "mypy --install-types" to install all missing stub packages)
__main__.py:6: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
__main__.py:11: error: Library stubs not installed for "tabulate" (or incompatible with Python 3.10)
__main__.py:11: note: Hint: "python3 -m pip install types-tabulate"
__main__.py:31: error: Name "FifoCostBasis" is not defined
__main__.py:31: error: Name "Sale" is not defined
__main__.py:36: error: Name "Optional" is not defined
__main__.py:36: note: Did you forget to import it from "typing"? (Suggestion: "from typing import Optional")
__main__.py:38: error: Name "true" is not defined
__main__.py:64: error: Name "Queue" is not defined
__main__.py:78: error: Name "Purchase" is not defined
__main__.py:81: error: Name "Sale" is not defined
__main__.py:87: error: Name "FifoCostBasis" is not defined
Found 21 errors in 2 files (checked 4 source files)
make: *** [mypy] Error 1

```

--- a/src/main.rs
+++ b/src/main.rs
@@ -16,7 +16,7 @@ mod cli;
 #[grammar = "mustfile.pest"]
 pub struct Pestfile;
 
-#[derive(Debug,Default)]
+#[derive(Debug,Default,Clone)]
 struct Task {
     target: String,
     _deps: Vec<String>,
@@ -24,9 +24,18 @@ struct Task {
     steps: Vec<String>
 }
 
-// TODO how to tell it to use the variant `task`?
-// TODO this needs to return the Result type
-fn execute_task(task: Task) -> Result<(), String> {
+fn execute_task(task: Task, all_tasks: &Vec<Task>) -> Result<(), String> {
+    // First execute all dependencies
+    for dep_name in &task._deps {
+        let dep_task = all_tasks.iter()
+            .find(|t| &t.target == dep_name)
+            .ok_or_else(|| format!("Dependency '{}' not found", dep_name))?;
+
+        println!("Executing dependency: {}", dep_name);
+        execute_task(dep_task.clone(), all_tasks)?;
+    }
+
+    // Then execute the main task steps
     for step in task.steps {
         println!("{}", step);
         let result = Command::new("sh")
@@ -46,10 +55,6 @@ fn execute_task(task: Task) -> Result<(), String> {
         }
     }
     Ok(())
-}
-
-fn sum_values(x: i16, y: i16) -> i32 {
-    (x as i32) + (y as i32)
 }
 
 fn build_task(taskentries: Pairs<Rule>) -> Task {
@@ -97,7 +102,7 @@ fn main() {
             }
         }
     } else {
-        let mut tasks = mustfile.map_while (|entry| -> Option<Task> {
+        let tasks: Vec<Task> = mustfile.map_while(|entry| -> Option<Task> {
             match entry.as_rule() {
                 Rule::task => {
                     let task = entry.into_inner();
@@ -106,14 +111,15 @@ fn main() {
                 Rule::EOI => None,
                 _ => unreachable!()
             }
-        });
+        }).collect();
 
         let task = match &args.command {
-            Some(command) => tasks.find(|t| &t.target == command),
-            None => tasks.next()
+            Some(command) => tasks.iter().find(|t| &t.target == command),
+            None => tasks.first()
         };
+
         if let Some(task) = task {
-            if let Err(e) = execute_task(task) {
+            if let Err(e) = execute_task(task.clone(), &tasks) {
                 eprintln!("Error: {}", e);
                 std::process::exit(1);
             }
@@ -176,9 +182,31 @@ mod tests {
 
         let task = file.next().unwrap();
         let task = build_task(task.into_inner());
-        let result = execute_task(task);
+        let result = execute_task(task, &vec![]);
         assert!(result.is_err());
         let error = result.unwrap_err();
         assert!(error.contains("Command failed with exit code: 1"));
+    }
+
+    #[test]
+    fn executes_dependencies_in_order() {
+        let testfile = fs::read_to_string("tests/fixtures/dependencies.must").expect("couldnt find test file");
+        let mut file = Pestfile::parse(Rule::must, &testfile)
+            .expect("unsuccessful parse");
+
+        let tasks: Vec<Task> = file.map_while(|entry| -> Option<Task> {
+            match entry.as_rule() {
+                Rule::task => {
+                    let task = entry.into_inner();
+                    Some(build_task(task))
+                },
+                Rule::EOI => None,
+                _ => unreachable!()
+            }
+        }).collect();
+
+        let main_task = tasks.iter().find(|t| t.target == "main").unwrap();
+        let result = execute_task(main_task.clone(), &tasks);
+        assert!(result.is_ok());
     }
 }

--- a/tests/fixtures/dependencies.must
+++ b/tests/fixtures/dependencies.must
@@ -1,0 +1,8 @@
+dep1:
+    echo "Running dep1"
+
+dep2: dep1
+    echo "Running dep2"
+
+main: dep1 dep2
+    echo "Running main"

